### PR TITLE
Ugly formatting of nested conditional types [TypeScript]

### PR DESCRIPTION
**Pull request by mmiszy**
Fixes https://github.com/prettier/prettier/issues/7940

My intention was to only affect nested conditional types in TypeScript and not to introduce any changes to how ternary expressions are formatted. This allows for a smoother transition, however, it might yield seemingly inconsistent results (see [this example](https://deploy-preview-7948--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKpuAOQEMBbOAHlwD5sBeAHSmz2zgA94oATAZ214wATgEsoAc2wB+bAxCDREudiRMWuNpwR9sUAK6kARnCHTZIfUZPLVzVhy47DECABs4xZjLnO3HqDZq9lo8-Ho8cABmYnDcZnLh3FEx3IF2Gg7a-ABi4WAwItDxIJF5BdBp6pqO-ACCQkLEaBSeaLTeIMQNTZXBNbp6rq7F+kO9GSE6vGhGbsXTs669chCGAFZw+XIA3ExMkFCC2AAa9EEadHQCwmKSHQq34-RXlsamHa-WICrnz9i+7k8xQB-iel2wiWSUFixUh0WhqW+tiq4NyUHyhS85lK6PKASRv3B9UaaGKXRJYJeg2GH2plIEM2cNPMCyZyxAqw2WxA2xAABoORg8bxkKByRAAO4ABS6CBFKGIrglTRFAsMjTAAGs4DAAMoYYhgW7IYR6OACgAWMFIrgA6haRPBeAawHBdXLHSIAG6OtDIcC8VUgMS8EwwKWNcSkYjISKK0MCta8dgAIQ12r1ZDgABkYrH4+aQEn2Lrbu4AIp6CDwfOuBMgA1CUNCf3oLC8MCiIX8huKGC2kTcGAW5AADgADAKMEIIKHbY0MP7p3Bm164D2AI5V+ARiAYeWdXgAWmhsViPaEcC3IkvEeIUZjSDjdcLodIIhNQjNAt4ZbglerdcnwLAUYGIQwByHEckAAJlAxoRFcW4AGEIFIaN-RXABWHs9FDXBwPlZ96y9M0AEkIlgXVOxEIVah4XV0HcWtQwAX1YoA)).

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by mmiszy** [comment]
Alright! Would you mind pasting the second example with your suggested formatting?

**Comment by thorn0** [comment]
The current formatting works great for that example.

**Comment by mmiszy** [comment]
Indeed. But what if you have both long and short types inside a nested conditional?

**Comment by thorn0** [comment]
Ideally, the new formatting should be used _only_ if each condition with its "true" branch fits on one line like in your original example, but as far as I understand Prettier's printer algorithm doesn't support such a trick (changing the formatting of previous lines if the current line exceeds the print width). This means we'll have to go for a less perfect solution: a heuristic that would check the AST and decide whether the new formatting would fit within the print width.

File: tests/typescript_conditional_types/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -151,33 +142,33 @@ type Foo3 =
...
+  (
+    ThingamabobberFactory extends AbstractThingamabobberFactory ?
+    GobbledygookProvider :
+    CompositeGobbledygookProvider
+  ) extends (
+    DoubleGobbledygookProvider extends MockGobbledygookProvider ?
+    MockThingamabobberFactory :
+    ThingamabobberFactory
+  ) ? UniqueDalgametreService :
+  CompositeZamazingoResolver;
```
Comment (ExE-Boss) [inline tests/typescript_conditional_types/__snapshots__/jsfmt.spec.js.snap:171]: These should remain unchanged.

**Comment by rattrayalex** [comment]
FYI https://github.com/prettier/prettier/issues/9561 contains proposals to use this style of formatting in TS Conditionals as well as, possibly, some nested ternaries. I wrote it before I saw this PR, but I might just close mine in favor of this.

Do you still plan to work on it? Would you like help handling cases where the clauses are long?

**Comment by kube** [comment]
This is something that hurts me since I started using Prettier on TypeScript libraries.

Usage of conditional types in this domain happens a lot more than in classical Applications, and I also think that the output is ugly.

It does not make any sense that the first condition is always next to `=`, like in:

```ts
export type SchemaToType<S extends Schema> = S extends SchemaLiteral
  ? SchemaLiteralToType<S>
  : S extends SchemaPrimitive
  ? SchemaPrimitiveToType<S>
  : S extends Schema.Record
  ? SchemaRecordToType<S>
  : S extends Schema.Array
  ? SchemaArrayToType<S>
  : S extends Schema.Intersection
  ? SchemaIntersectionToType<S>
  : S extends Schema.Union
  ? SchemaUnionToType<S>
  : S;
````

Just allowing to have the first condition on the next line (maybe through an config option?), would be a big win IMO.

```ts
export type SchemaToType<S extends Schema> =
  S extends SchemaLiteral
  ? SchemaLiteralToType<S>
  : S extends SchemaPrimitive
  ? SchemaPrimitiveToType<S>
  : S extends Schema.Record
  ? SchemaRecordToType<S>
  : S extends Schema.Array
  ? SchemaArrayToType<S>
  : S extends Schema.Intersection
  ? SchemaIntersectionToType<S>
  : S extends Schema.Union
  ? SchemaUnionToType<S>
  : S;
````

**Comment by kube** [comment]
Any news on this @mmiszy @thorn0 ?
Can I help somehow ?

**Comment by thorn0** [comment]
@kube I've opened a separate issue for that

**Comment by thorn0** [comment]
Doing a clean up of stale PRs, closing this. It's linked to the issue so we can revive it if necessary.

**Pull request closed without merging**